### PR TITLE
Convert elliptic boundary conditions to use factory map

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

List the available boundary conditions in `Metavariables::factory_creation` and remove all the registration code from the individual classes.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp [hunk 1]
```diff
@@ -12,11 +12,31 @@
...
 #include "Utilities/TMPL.hpp"
 
 namespace elliptic {
+
+namespace detail {
+// Return the `BoundaryConditionClasses`, or get the list of derived classes
+// from `Metavariables::factory_creation` if `BoundaryConditionClasses` is
+// empty.
+template <typename Base, typename BoundaryConditionClasses, typename DbTagsList>
+struct GetBoundaryConditionClasses {
```
Comment (wthrowe) [inline src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp:27]: I'm not quite sure what this is for, but my guess based on the rest of the patch is that it is to allow different boundary conditions on the linear and non-linear solves.  Would templating the base class on the intended use and then specifying both lists in the metavariables work?  I don't know where the alternate list comes from, so it might not.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp (hunk 1 above)
Comment (nilsvu) [reply src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp:27]: The main use case for this option to bypass the metavars is for the `MinusLaplacian` subdomain preconditioner. It's in the diff of this PR. It uses a Poisson system with a particular boundary condition to approximate the equations, so I pass those boundary conditions down to the operator. I think it works quite well, and it also comes in handy in the tests 🤷‍♂️

Edit: FYI, the linear vs. non-linear boundary conditions are handled separately. They use the same class, but which member function is called is toggled by the `Linearized` template parameter.

Comment (wthrowe) [reply src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp:27]: OK, I think I see.  Is the following correct?

This feature is used by `MinusLaplacian`, which always uses a `Robin` boundary condition.  That condition class is stored in the `MinusLaplacian` class.  It doesn't depend on anything, so it isn't checkpointed and is just constructed whenever `MinusLaplacian` is.  When `MinusLaplacian` calls into the code to apply the boundary condition, it passes that in and, as a template parameter, the information that the class is guaranteed to be `Robin`, not whatever the main solve is using.

If so, that seems fine.  (And use in tests is also fine.)

Comment (nilsvu) [reply src/Elliptic/BoundaryConditions/ApplyBoundaryCondition.hpp:27]: That's exactly right

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsvu** [comment]
Rebased on latest develop

File: src/Elliptic/Systems/Poisson/BoundaryConditions/Factory.hpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
```
Comment (yoonso0-0) [inline src/Elliptic/Systems/Poisson/BoundaryConditions/Factory.hpp]: [nitpik] not used?

Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/BoundaryConditions/Factory.hpp]: deleted this line 👍

File: src/Elliptic/Systems/Xcts/BoundaryConditions/Factory.hpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
```
Comment (yoonso0-0) [inline src/Elliptic/Systems/Xcts/BoundaryConditions/Factory.hpp]: [nitpik] not used

File: src/Elliptic/Systems/Poisson/BoundaryConditions/Robin.hpp [hunk 4]
```diff
@@ -71,74 +89,29 @@ struct RobinImpl {
...
-/// Impose Robin boundary conditions \f$a u + b n_i \nabla^i u = c\f$. The
-/// boundary condition is imposed as Neumann-type (i.e. on \f$n_i \nabla^i u\f$)
-/// if \f$|b| > 0\f$ and as Dirichlet-type (i.e. on \f$u\f$) if \f$b = 0\f$.
-template <size_t Dim, typename Registrars = tmpl::list<Registrars::Robin<Dim>>>
-class Robin
-    : public elliptic::BoundaryConditions::BoundaryCondition<Dim, Registrars>,
-      public detail::RobinImpl {
- private:
-  using Base = elliptic::BoundaryConditions::BoundaryCondition<Dim, Registrars>;
+bool operator==(const Robin<Dim>& lhs, const Robin<Dim>& rhs) {
```
Comment (yoonso0-0) [inline src/Elliptic/Systems/Poisson/BoundaryConditions/Robin.hpp]: [nitpik] I think for consistency with other parts of this PR it'd be good to move definition of operators `==` `!=` and PUP ID (line 113) part to .cpp file.

File: src/Elliptic/Systems/Xcts/BoundaryConditions/ApparentHorizon.hpp [hunk 5]
```diff
@@ -230,134 +309,16 @@ struct ApparentHorizonImpl {
...
-    Base::pup(p);
-    detail::ApparentHorizonImpl<ConformalGeometry>::pup(p);
-  }
-};
-
-/// \cond
-template <Xcts::Geometry ConformalGeometry, typename Registrars>
-PUP::able::PUP_ID ApparentHorizon<ConformalGeometry, Registrars>::my_PUP_ID =
-    0;  // NOLINT
+PUP::able::PUP_ID ApparentHorizon<ConformalGeometry>::my_PUP_ID = 0;  // NOLINT
```
Comment (yoonso0-0) [inline src/Elliptic/Systems/Xcts/BoundaryConditions/ApparentHorizon.hpp]: Same here (to .cpp)

**Review by yoonso0-0 — COMMENTED** [review]
Yep, I've finished taking a look at this as far as I could comprehend :)

File: src/Elliptic/Systems/Poisson/BoundaryConditions/Robin.hpp (hunk 4 above)
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/BoundaryConditions/Robin.hpp]: done and pushed in a fixup 👍

**Review by yoonso0-0 — DISMISSED** [review]
LGTM! Please squash :)

**Review by wthrowe — APPROVED** [review]

**Comment by nilsvu** [comment]
Thanks for reviewing @wthrowe and @yoonso0-0!
